### DISCUSSION TITLE
Partial Combinatorial Generation

### DISCUSSION THREAD

**Discussion by DrJimmyBrungus** [adieyal/sd-dynamic-prompts#762] [Q&A] [answered]
Let's say I have the following:

```
{cat | dog}
{black fur | brown fur}
```

Using combinatorial generation and a fixed seed, I would get 4 results:
A black cat, a brown cat, a black dog and a brown dog. All good so far.

But now let's say I also want to pick a random artist from a wildcard file containing hundreds of artists:

```
{cat | dog}
{black fur | brown fur}
by __artist__
```

Using combinatorial generation here will result in hundreds and hundreds of generations. 
I can try setting the max generations to 4, but the problem here is it seems to want to first go through the list of artists alphabetically, using the same combo of dog / cat / black / brown for each, so the only variation I am getting is the artist.

Is it possible to get this kind of behavour? Note that there are also two ways this could go - either a random artist is selected for each of the 4 generations, e.g.

black cat, by ??artist??
brown cat, by ??artist??
black dog, by ??artist??
brown dog, by ??artist??

Or the random artist is picked first and the same one used each time

black cat, by Bob Ross
brown cat, by Bob Ross
black dog, by Bob Ross
brown dog, by Bob Ross

Ideally I'd be able to choose either option as both have their uses.

**Comment by akx** [comment, marked as answer]
Based on [the samplers syntax](https://github.com/adieyal/sd-dynamic-prompts/blob/main/docs/SYNTAX.md#samplers) you should be able to do
```
{cat | dog}
{black fur | brown fur}
by __~artist__
```
(tilde within the wildcard wrap) to override the sampler for that command.

**Reply by DrJimmyBrungus** [reply]
Nice, that worked perfectly to give me a random artist each time for all of the 4 combos.

From that I figured out how to get the all possible combos using the same random artist for each, by using variables:

```
{cat | dog}
{black fur | brown fur}
${artist=!__~testartist__}
by ${artist}
```

The ! make it so the artist is decided when you hit generate instead of every time it is called, and like before, the ~ picks that artist randomly from the wildcard list instead of cycling from the start.

**Comment by DrJimmyBrungus** [comment]
I'm thinking this is impossible, but any idea how I might use combinatorial wildcards together with Prompt S/R on XYZ plot?

The idea is to pick from the list of artists from an artist.txt wildcard file, and actually show that artist's name on the XYZ plot. 
So far all I can get is the text `"__artist__"` showing on the plot.

**Reply by akx** [reply]
Not sure that's easily possible for two reasons:

* we don't currently e.g. expose variable values or whatnot into the infotext
* there's no option I know of to have an arbitrary label (that'd be something read from infotext) in the batch grid/XYZ plot anyway

With a bit of code, you could generate e.g. a HTML page out of generated images that'd show the prompts next to them.

**Comment by DrJimmyBrungus** [comment]
I thought as much. I guess prompt S/R is not really the right tool for the job anyway, as it doesn't even make sense to search and replace `__artist__` with.. what? `__artist__` again? 

I guess the best I can do is just convert the wildcard list of artists into a comma separated line of artists, and use that instead in the S/R plot. Just means I won't be able to have any randomness, but this would'nt be needed anyway if I just want to compare all artists.
